pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * The name of Imperial College London or that of other
+#       contributors may not be used to endorse or promote products
+#       derived from this software without specific prior written
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTERS
+# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""OP2 backend for fusion and tiling of ``ParLoops``."""
+
+from contextlib import contextmanager
+from collections import OrderedDict
+from copy import deepcopy as dcopy, copy as scopy
+import os
+
+from base import *
+import base
+import compilation
+import host
+from backends import _make_object
+from caching import Cached
+from profiling import lineprof, timed_region, profile
+from logger import warning, info as log_info
+from mpi import collective
+from configuration import configuration
+from utils import flatten, strip, as_tuple
+
+import coffee
+from coffee import base as ast
+from coffee.plan import ASTKernel
+from coffee.utils import visit as ast_visit, ast_c_make_alias as ast_make_alias
+
+import slope_python as slope
+
+
+class Arg(host.Arg):
+
+    @staticmethod
+    def specialize(args, gtl_map, loop_id):
+        """Given ``args``, instances of some :class:`fusion.Arg` superclass,
+        create and return specialized :class:`fusion.Arg` objects.
+
+        :param args: either a single :class:`host.Arg` object or an iterator
+                     (accepted: list, tuple) of :class:`host.Arg` objects.
+        :gtl_map: a dict associating global maps' names to local maps' c_names.
+        :param loop_id: indicates the position of the args` loop in the loop
+                        chain
+        """
+
+        def convert(arg, gtl_map, loop_id):
+            # Retrive local maps
+            maps = as_tuple(arg.map, Map)
+            c_local_maps = [None]*len(maps)
+            for i, map in enumerate(maps):
+                c_local_maps[i] = [None]*len(map)
+                for j, m in enumerate(map):
+                    c_local_maps[i][j] = gtl_map["%s%d_%d" % (m.name, i, j)]
```

Comment:
Ah right, yes. I put it as a string because later one uses that string, so sooner or later the conversion will happen anyway. But I can postpone the conversion if you feel it'd be much cleaner